{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Week 4.ipynb","provenance":[],"mount_file_id":"1YfF87SrqYLElRWxIXv8ykUveEJ-r3VyU","authorship_tag":"ABX9TyP6JbMvQyOFApMi7zksdItH"},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"hN9A7vhWHYEJ","colab_type":"text"},"source":["1. **GetProfileMatrix(motifs)**: Given a list of motifs, returns a dictionary representing the profile matrix of those motifs.\n","2. **GetProfMostProbKmer(genome, k, prof_dict)**: Given a genome, k-mer size, and a profile matrix (prof_dict), returns the first most probable k-mer from the genome. Uses pseudocounts.\n","3. **GetProfMostProbMotifs(DNA, prof_dict)**: Given a DNA, k-mer size, and a profile matrix (prof_dict), returns the list of first most probable k-mer in the genomes from the DNA. Uses GetProfMostProbKmer(genome, k, prof_dict).\n","4. **GetMotifsScore(motifs)**: Given a list of motifs, returns the Score from those motifs. The score is calculated by counting the least frequent sites from every column of the motif matrix.\n","5. **GetRandomizedMotifSearch(DNA, k, t)**: Given DNA, kmer size k and the number of genomes in the DNA t, returns a list of motifs. From randomly selected motifs from the DNA, it uses these motifs to generate profile matrix, and from this profile matrix, it again generates motifs using GetProfMostProbKmer. This process continues until suitable score of motifs is reached. Remember to run it multiple times from random start to get better motifs.\n","6. **GetProfProbGeneList(genome, k, prof_dict)**: Given genome, k and a profile matrix prof_dict, returns a list of probability of the all the k-mers in that genome.\n","7. **GetGibbsSampler(DNA, k, t, N)**: Given DNA, k, number of genes in the DNA t, and the number of times this Gibbs Sample would be run N, it returns a list of motifs. From randomly selected motifs, it generates profile matrix discarding one randomly selected motifs. It uses this profile matrix in GetProfProbGeneList to generate a probability list of that discarded gene, and uses this list of probabilites to choose one k-mer biasly. This k-mer is put back to the list of motifs and its score is evaluated. This process continues N times. Finally it returns the best_motifs it have found within these N iterations."]},{"cell_type":"code","metadata":{"id":"pwEuboXj9gdd","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":54},"executionInfo":{"status":"ok","timestamp":1594655198421,"user_tz":-360,"elapsed":1210,"user":{"displayName":"Md. Hasin Abrar","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GjtGDQ2wwOA1MtqidTjeLaqWyiZT1rvuQoN_MT7=s64","userId":"07235195413912240428"}},"outputId":"88e209d6-ff94-4b7a-dac9-ee5f0f25f522"},"source":["from google.colab import drive\n","drive.mount('/content/drive')\n","main_directory = '/content/drive/My Drive/Colab Notebooks/Bioinformatics Code Challenges/Data/'"],"execution_count":null,"outputs":[{"output_type":"stream","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"jZM8DUPyIP6I","colab_type":"code","colab":{}},"source":["def GetProfileMatrix(motifs):\n","  Site2Int = {s:i for i,s in enumerate('ACGT')}\n","  Int2Site = {i:s for i,s in enumerate('ACGT')}\n","  prof_dict = {s:[] for s in 'ACGT'}\n","  l = len(motifs) + 1\n","  k = len(motifs[0])\n","  for i in range(k):\n","    sites = [1.0 for _ in range(4)]\n","    for motif in motifs:\n","      sites[Site2Int[motif[i]]] += 1\n","    for i in range(4):\n","      sites[i] /= l\n","      prof_dict[Int2Site[i]].append(sites[i])\n","  return prof_dict\n","\n","def GetProfMostProbKmer(genome, k, prof_dict):\n","  _max = -1\n","  most_prob_kmer = ''\n","  for i in range(len(genome)-k+1):\n","    # print(type(k))\n","    sub_kmer = genome[i:i+k]\n","    val = 1\n","    for j,km in enumerate(sub_kmer):\n","      val *= prof_dict[km][j]\n","    if val > _max:\n","      _max = val\n","      most_prob_kmer = sub_kmer\n","  return most_prob_kmer\n","\n","def GetProfMostProbMotifs(DNA, prof_dict):\n","  k = len(prof_dict['A'])\n","  motifs = []\n","  for genome in DNA:\n","    motif = GetProfMostProbKmer(genome, k, prof_dict)\n","    motifs.append(motif)\n","  return motifs\n","\n","def GetMotifsScore(motifs):\n","  Site2Int = {s:i for i,s in enumerate('ACGT')}\n","  Int2Site = {i:s for i,s in enumerate('ACGT')}\n","  prof_dict = {s:[] for s in 'ACGT'}\n","  l = len(motifs)\n","  k = len(motifs[0])\n","  score = 0\n","  for i in range(k):\n","    sites = [0 for _ in range(4)]\n","    for motif in motifs:\n","      sites[Site2Int[motif[i]]] += 1\n","    _max = max(sites)\n","    score += (l-_max)\n","  return score\n","\n","# RandomizedMotifSearch\n","def GetRandomizedMotifSearch(DNA, k, t):\n","  import random\n","  best_motifs = []\n","  motifs = []\n","  for m in DNA:\n","    i = random.randint(0,len(m)-k)\n","    best_motifs.append(m[i:i+k])\n","    motifs.append(m[i:i+k])\n","  best_score = GetMotifsScore(best_motifs)\n","  while True:\n","    prof_dict = GetProfileMatrix(motifs)\n","    motifs = GetProfMostProbMotifs(DNA, prof_dict)\n","    motifs_score = GetMotifsScore(motifs)\n","    if motifs_score < best_score:\n","      best_score = motifs_score\n","      best_motifs = motifs\n","    else:\n","      break\n","  return best_motifs\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"CWZlsw8RMIKD","colab_type":"code","colab":{}},"source":["def CheckGetRandomizedMotifSearch():\n","  file = open(main_directory+\"Week4/rms1.txt\", \"r\")\n","  lines = file.readlines()\n","  k = 15\n","  t = 20\n","  DNA = []\n","  for i in range(1,len(lines)):\n","    DNA.append(lines[i].strip())\n","  best_motifs = GetRandomizedMotifSearch(DNA,k,t)\n","  best_score = GetMotifsScore(best_motifs)\n","  for i in range(1000):\n","    motifs = GetRandomizedMotifSearch(DNA,k,t)\n","    motifs_score = GetMotifsScore(motifs)\n","    if motifs_score < best_score:\n","      best_score = motifs_score\n","      best_motifs = motifs\n","  for m in best_motifs:\n","    print(m)\n","  \n","CheckGetRandomizedMotifSearch()"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"xrdc_qD6Thyq","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":34},"executionInfo":{"status":"ok","timestamp":1594625802443,"user_tz":-360,"elapsed":968,"user":{"displayName":"Md. Hasin Abrar","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GjtGDQ2wwOA1MtqidTjeLaqWyiZT1rvuQoN_MT7=s64","userId":"07235195413912240428"}},"outputId":"18411224-cb09-448f-a004-752b6868e5bc"},"source":["x =((585**9/ 586**10)*10 + (585/586)**10 )\n","x = 1 - x\n","print(x)"],"execution_count":null,"outputs":[{"output_type":"stream","text":["0.00012985670567622343\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"I2d09piauIWp","colab_type":"code","colab":{}},"source":["def GetProfProbGeneList(genome, k, prof_dict):\n","  prof_prob_list = []\n","  for i in range(len(genome)-k+1):\n","    sub_kmer = genome[i:i+k]\n","    val = 1\n","    for j,km in enumerate(sub_kmer):\n","      val *= prof_dict[km][j]\n","    prof_prob_list.append(val)\n","  return prof_prob_list\n","\n","# GibbsSampler(Dna, k, t, N)\n","def GetGibbsSampler(DNA, k, t, N):\n","  import random\n","  best_motifs = []\n","  motifs = []\n","  for genome in DNA:\n","    i = random.randint(0,len(genome)-k)\n","    best_motifs.append(genome[i:i+k])\n","    motifs.append(genome[i:i+k])\n","  best_score = GetMotifsScore(best_motifs)\n","  for j in range(N):\n","    i = random.randint(0,t-1)\n","    temp_motifs = motifs[0:i] + motifs[i+1:]\n","    prof_dict = GetProfileMatrix(temp_motifs)\n","    genome = DNA[i]\n","    prof_prob_list = GetProfProbGeneList(genome,k,prof_dict)\n","    biased_choice = random.choices(range(len(prof_prob_list)), \n","                                   weights=prof_prob_list, k = 1)\n","    motifs[i] = DNA[i][biased_choice[0]:biased_choice[0]+k]\n","    motifs_score = GetMotifsScore(motifs)\n","    if motifs_score < best_score:\n","      best_score = motifs_score\n","      best_motifs = motifs[:]\n","  return best_motifs\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"d2r9mR8S2GMW","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":353},"executionInfo":{"status":"ok","timestamp":1594657707562,"user_tz":-360,"elapsed":28897,"user":{"displayName":"Md. Hasin Abrar","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GjtGDQ2wwOA1MtqidTjeLaqWyiZT1rvuQoN_MT7=s64","userId":"07235195413912240428"}},"outputId":"9c14291e-5a0b-4cbf-b9ec-b942a9ff5394"},"source":["def CheckGetGibbsSampler():\n","  file = open(main_directory+\"Week4/gs2.txt\", \"r\")\n","  lines = file.readlines()\n","  k = 15\n","  t = 20\n","  N = 2000\n","  DNA = []\n","  for i in range(1, len(lines)):\n","    DNA.append(lines[i].strip())\n","  best_motifs = GetGibbsSampler(DNA, k, t, N)\n","  best_score = GetMotifsScore(best_motifs)\n","  for i in range(20):\n","    motifs = GetGibbsSampler(DNA, k, t, N)\n","    motifs_score = GetMotifsScore(motifs)\n","    if motifs_score < best_score:\n","      best_score = motifs_score\n","      best_motifs = motifs\n","  for m in best_motifs:\n","    print(m)\n","\n","CheckGetGibbsSampler()"],"execution_count":null,"outputs":[{"output_type":"stream","text":["TACAACAGGGGGGTT\n","CAGCTGGCGGTGGTG\n","AAGAACGCGGTGGGC\n","CAGAACGCGTCAGTG\n","CAGAGTACGGTGGTG\n","CAGAACGTCATGGTG\n","CACCTCGCGGTGGTG\n","CAGAACGCGGCTATG\n","CAGAAGTGGGTGGTG\n","TCGAACGCGGTGGTA\n","CAGAATCGGGTGGTG\n","CAGGGGGCGGTGGTG\n","TGAAACGCGGTGGTG\n","CAGAACCACGTGGTG\n","CAGAACTGCGTGGTG\n","CAGAACGCGGTGTGT\n","CGCGACGCGGTGGTG\n","CAGAGACCGGTGGTG\n","CAGAACGCTCGGGTG\n","CAGAACGCGGTATAG\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"yYrlkx0iu1Ly","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":34},"executionInfo":{"status":"ok","timestamp":1594654944764,"user_tz":-360,"elapsed":1182,"user":{"displayName":"Md. Hasin Abrar","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GjtGDQ2wwOA1MtqidTjeLaqWyiZT1rvuQoN_MT7=s64","userId":"07235195413912240428"}},"outputId":"053cac2f-98a3-4d5b-bff7-70ca3263cc77"},"source":["x = [i for i in range(3)]\n","y = x[:]\n","x[0] = 10\n","y[2] = 20\n","print(x,y)"],"execution_count":null,"outputs":[{"output_type":"stream","text":["[10, 1, 2] [0, 1, 20]\n"],"name":"stdout"}]}]}